### DISCUSSION TITLE
Narrow completion suggestions to __all__ value

### DISCUSSION THREAD

**Discussion by m-novikov** [microsoft/pylance-release#3709] [Enhancement]
**Is your feature request related to a problem? Please describe.**

Some modules import a lot of symbols into their namespace. 
Auto-completion on such modules lists own symbols as well as all imported ones, which can be inconvenient.

Let's say I have a module `a.py` containing following
```python
from typing import List

__all__ = ["foo"]

def foo() -> List[str]:
    return ["a"]
```
Currently autocomplete on `a.` gives me `a.List` and `a.foo` as possible options.

**Describe the solution you'd like**

When triggering auto complete on such module I would like to see only names listed in `__all__` in suggestions + common module attributes such as `__file__`, `__doc__`

**Additional context**
For modules that don't specify `__all__` all names should be listed as before.

**Comment by erictraut** [comment]
I don't think this matches the expectations of most users. The `__all__` symbol is not meant to be a comprehensive list of exported symbols. Rather, it's a list of symbols that are included in a wildcard import.

Since this isn't a core type checking issue, I'm going to transfer it to the pylance-release project and let the pylance team decide how to proceed.

**Comment by erictraut** [comment]
@m-novikov, is the module part of a "py.typed" library? If so, then we could limit suggestions only to those symbols that are considered part of the public interface for the library. For modules that are not part of a "py.typed" library (or type stubs), we don't have a good way of knowing which symbols are meant to be public and which are not. Python doesn't have any keywords like `public` and `private`, so we generally need to assume that the module author meant for all symbols to be public. See [this documentation](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) for more details.

**Comment by m-novikov** [comment]
It's not a part of `py.typed`. Maybe it makes sense to rank "private" symbols lower in suggestions?

**Comment by pukhov** [comment]
Are there any updates on this?

+1 for first showing all attributes defined in `__all__`, if present, and then show all other "more private like" functions/attributes. I understand it is not what `__all__` means, but it kind of does mean this.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by joelostblom** [comment]
As a library maintainer I find it useful to be able to control what shows up in tab completion to improve the user experience. For example, as a library grows I might want to avoid that the top level API becomes too cluttered by moving some top-level modules into sub-modules, but at the same time I don't want to break existing code.

In JupyterLab, I am able to do this by modifying `__all__` to remove items from the top-level tab-completion while still importing the at top level to ensure backwards-compatibility. I would like to provide the same experience to users of VS Code, but haven't found a way of doing so yet. What's an equivalent mechanism that I can use in pylance to control what is shown in the modules' top-level tab completion?

From reading the [docs linked above](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface), I see that I can make modules private, but I don't understand how to make individual classes and functions now show up in tab completion while still allowing them to be accessible from the top level if typed out. From this comment https://github.com/microsoft/pylance-release/issues/2572#issuecomment-1100566151 I see that I can use `__all__` to include individual modules, and it would be convenient with something equivalent for *excluding* modules instead.

**Reply by erictraut** [reply]
Is your library a typed library? In other words, does it include inlined types and indicate as such by supplying a "py.typed" marker file? If not, there's relatively little you can do to influence which symbols show up in completion suggestions.

If your library is already typed, then you're most of the way there. You can control which symbols are publicly exported from each module. Pylance should show only the public exports in completion suggestions. The docs you cited above indicate how to specify whether a symbol is considered public or private.

For example, let's say that you have a symbol declared in submodule `foo.py` called Foo. You can import this symbol in your top-level module by doing the following:
```python
# __init__.py
from .foo import Foo
```

In this case, the symbol `Foo` is considered "private" for the top-level module and should therefore not appear as a completion suggestion. If someone happens to import the symbol as `from <lib> import Foo`, it will work at runtime (since Python has no runtime distinction of "private" and "public"), but Pylance will report such an import as an error if the user has enabled `reportPrivateImportUsage`.

If you want `Foo` to be explicitly re-exported from your top-level module as a public symbol, you'd need to do one of the following:
1. Add it to the `__all__` list in your top-level module.
2. Use a redundant form of import, as in `from .foo import Foo as Foo`.

**Reply by joelostblom** [reply]
Thanks for the quick and detailed reply @erictraut ! The library I have in mind is Altair, where we unfortunately have had some struggles adding proper typing support. There have been an initial attempt in https://github.com/altair-viz/altair/pull/2614, but currently we lack the time and to some degree expertise to progress. From your description it sounds like this is another reason to try to make the addition of types a priority, since it is the only way to enable control over the top-level completion menu.

**Reply by binste** [reply]
This is very insightful, thank you @erictraut! Even though Altair can't be considered type complete yet, I wanted to try if we can just restrict top-level autocompletion and followed your suggestions. However, completion suggestions still show private attributes, not just public exports. I summarised the current and expected behavior with screenshots in this short draft PR https://github.com/altair-viz/altair/pull/2927, if you could take a quick look it would be much appreciated!

**Reply by erictraut** [reply]
@binste, I took a quick look at your PR, and it looks correct to me. If pylance is displaying all symbols in the completion list, even those that are considered private symbols in a "py.typed" library, then I'd consider that a bug, and the pylance team should look into it.

**Reply by binste** [reply]
Thank you! I opened https://github.com/microsoft/pylance-release/issues/4059.
